### ISSUE TITLE
build cloud-hosted multiple-validator setup tool

### ISSUE THREAD

**Issue by warner** [Agoric/cosmic-swingset#9]
@michaelfig and I sketched out a plan for setting up multiple validators and configuring them to talk to each other. The idea is to launch a docker image on each of several providers, and each image will run `ag-chain-cosmos init`, then listen on a little webserver on a predefined port for a coordinator process to ping it. It will have two endpoints. The first will return the new agcc public keys. The coordinator will fetch these from all nodes and use them to build the combined genesis.json. It will also collect the hostname/ipaddr/port of all nodes and build a shared addressbook so they can all find each other. It will also collect the initial vat code (and bootstrap.js).

The coordinator will then send the whole bundle to the second endpoint, which will modify the agcc setup files with the bundle's contents. Then the webserver will shut down, and it will switch the image into runtime mode, where it runs `agcc start`. If we bounce the image after setup, it should go directly into `agcc start` and skip the setup phase.

There's a race here, of course, which we'll fix eventually by baking some keypair or something into the image and signing the setup message.

We need to build the coordinator tool as well: it should take a list of ipaddr/port (of the newly-launched agcc images waiting to be configured) and a local directory full of vat code (part of which is a bootstrap file that will contain the pubkey of the controller solo node), and it should contact all the chain nodes and configure them and set them running.

Maybe this coordinator tool should also take instructions on how to contact the PaaS APIs and launch the hosts (giving it API keys and whatever). But maybe not.. decouple the creation of the chain hosts from their configuration.

**Comment by michaelfig** [comment]
I've been following through [tendermint testnet setup instructions](https://tendermint.com/docs/networks/terraform-and-ansible.html) which use Terraform and Ansible, and they already do the setup using an installed SSH public key (duh!), and phase it so that there are no races.

Unfortunately, the instructions are broken in a few different ways, but fixing them and adapting them to `ag-chain-cosmos` is what I'm working on now.  Abstracting this model so that it can be used on multiple providers is my next goal, once I get 5 validators running in DigitalOcean.

Sorry, should have let you know about that before you took the time to make this detailed writeup.

**Comment by michaelfig** [comment]
I will be creating a Docker container that has all the tools needed to do the phases of the deployment to all supported clouds, then provide a Node.js entrypoint to access all the subcommands from a convenient shell script that just depends on Docker.

**Comment by warner** [comment]
Nah, no worries, I just wanted to nail down what we talked about so I could keep things straight in my head.

**Comment by warner** [comment]
in the old repo. this was cosmic-swingset issue 9
